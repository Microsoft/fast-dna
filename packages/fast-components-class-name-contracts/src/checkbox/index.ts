/**
 * The classname contract for the checkbox component
 */
export interface ICheckboxClassNameContract {
    checkbox: string;
}
