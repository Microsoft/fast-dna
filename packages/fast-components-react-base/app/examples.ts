import ButtonExamples from "../src/button/examples.data";
export { ButtonExamples };
<<<<<<< HEAD

import DividerExamples from "../src/divider/examples.data";
export { DividerExamples };

import ImageExamples from "../src/image/examples.data";
export {ImageExamples};

import ToggleExamples from "../src/toggle/examples.data";
export { ToggleExamples };

import TypographyExamples from "../src/typography/examples.data";
export { TypographyExamples };

import HyperlinkExamples from "../src/hyperlink/examples.data";
export { HyperlinkExamples };
=======
export {ButtonExamples};

import HypertextExamples from "../src/hypertext/examples.data";
export {HypertextExamples};
>>>>>>> renamed all to hypertext from hyperlink
