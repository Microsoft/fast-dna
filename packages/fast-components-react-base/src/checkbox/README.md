# Checkbox
The [checkbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox) component is an inout form element used to select or deselect items in a list. The list can be a single item with *checkbox* or multiple items, each with their own *checkbox*. *Checkboxes* have three visual states: checked, unchecked, and indeterminate.

## Accessibility
Consult the current [W3C WAI-ARIA documentation](https://www.w3.org/TR/wai-aria-practices/#checkbox) for best practices.