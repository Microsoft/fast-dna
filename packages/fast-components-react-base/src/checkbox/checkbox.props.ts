import * as React from "react";
import { ICheckboxClassNameContract, IManagedClasses } from "@microsoft/fast-components-class-name-contracts";

export interface ICheckboxHandledProps {
    /**
     * The checked state
     */
    checked?: boolean;

    /**
     * The textual content
     */
    text?: string;

    /**
     * The indeterminate option
     */
    indeterminate?: boolean;

    /**
     * The onChange event handler
     */
    onChange?: CheckboxOnChange;
}

export type CheckboxOnChange = (event?: React.ChangeEvent<HTMLElement>) => void;
export interface ICheckboxUnhandledProps extends React.AllHTMLAttributes<HTMLElement> {}
export interface ICheckboxManagedClasses extends IManagedClasses<ICheckboxClassNameContract> {}
export type CheckboxProps = ICheckboxHandledProps & ICheckboxUnhandledProps & ICheckboxManagedClasses;
