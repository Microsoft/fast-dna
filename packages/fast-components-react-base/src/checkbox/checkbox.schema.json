{
    "$schema": "http://json-schema.org/schema#",
    "title": "Checkbox",
    "description": "A checkbox component's schema definition.",
    "type": "object",
    "id": "checkbox",
    "children": true,
    "properties": {
        "checked": {
            "title": "Checked",
            "type": "boolean"
        },
        "disabled": {
            "title": "Disabled",
            "type": "boolean"
        },
        "indeterminate": {
            "title": "Indeterminate",
            "type": "boolean"
        }
    }
}
