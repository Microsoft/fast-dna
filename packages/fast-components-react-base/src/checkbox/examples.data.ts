import { ICategoryItemProps } from "@microsoft/fast-development-site-react";
import { ISnapshotTestSuite } from "@microsoft/fast-jest-snapshots-react";
import { IManagedClasses } from "@microsoft/fast-components-class-name-contracts";
import Checkbox, { ICheckboxHandledProps, ICheckboxManagedClasses, ICheckboxUnhandledProps } from "./checkbox";
import * as React from "react";

const examples: ISnapshotTestSuite<ICheckboxHandledProps & ICheckboxManagedClasses> = {
    name: "checkbox",
    component: Checkbox,
    data: [
        {
            managedClasses: {
                checkbox: "checkbox",
                checkbox_input: "checkbox-input",
                checkbox_label: "checkbox-label"
            }
        }
    ]
};

export default examples;
