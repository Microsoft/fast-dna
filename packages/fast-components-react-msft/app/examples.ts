import ButtonExamples from "../src/button/examples.data";
export {ButtonExamples};

import MediaExamples from "../src/media/examples.data";

export {MediaExamples};

import DividerExamples from "../src/divider/examples.data";
export {DividerExamples};
