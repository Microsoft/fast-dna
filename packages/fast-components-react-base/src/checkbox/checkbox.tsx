import * as React from "react";
import * as ReactDOM from "react-dom";
import Foundation, {HandledProps} from "../foundation";
import {ICheckboxHandledProps, ICheckboxItem, ICheckboxManagedClasses, ICheckboxUnhandledProps} from "./checkbox.props";
import {ICheckboxClassNameContract, IManagedClasses} from "@microsoft/fast-components-class-name-contracts";

/* tslint:disable-next-line */
class Checkbox extends Foundation<ICheckboxHandledProps & ICheckboxManagedClasses, ICheckboxUnhandledProps, {}> {
    protected handledProps: HandledProps<ICheckboxHandledProps & IManagedClasses<ICheckboxClassNameContract>> = {
        managedClasses: void 0,
        items: void(0)
    };

    /**
     * Generate HTML attributes based on props
     */
    public generateAttributes(item) {
        let attributes = {};

        if (item.disabled) {
            attributes["disabled"] = "disabled";
        }

        if (item.indeterminate) {
            // this is here to maintain shape
            attributes["data-js-checkbox"] = "indeterminate";
        }

        attributes["id"] = item.id;
        attributes["name"] = item.name;
        attributes["value"] = item.value;
        attributes["aria-label"] = item.text;

        return attributes;
    }

    /**
     * Renders a checkbox item
     */
    public renderCheckboxItem(item: ICheckboxItem, index: number) {
        return (
            <label key={item.id + index} className="c-label">
                <input
                    type="checkbox"
                    {...this.generateAttributes(item)}
                    onChange={item.onChange}
                    ref={this.setRef("items", index)}
                />
                <span dangerouslySetInnerHTML={{ __html: item.text }} />
            </label>
        );
    }

    /**
     * Renders checkbox items
     */
    public renderCheckboxItems() {
        return this.props.items.map((item, index) => {
            return this.renderCheckboxItem(item, index);
        });
    }

    /**
     * Renders the component
     */
    public render(): React.ReactElement<HTMLInputElement> {
        return (
            <div
                {...this.unhandledProps()}
                className={this.generateClassNames()}
            >
                {this.renderCheckboxItems()}
            </div>
        );
    }

    /**
     * Generates class names
     */
    protected generateClassNames(): string {
        return super.generateClassNames(this.props.managedClasses.checkbox);
    }
}

export default Checkbox;
export * from "./checkbox.props";
export {ICheckboxClassNameContract};
