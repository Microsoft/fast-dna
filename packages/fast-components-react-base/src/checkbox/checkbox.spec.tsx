import * as React from "react";
import * as ShallowRenderer from "react-test-renderer/shallow";
import * as Adapter from "enzyme-adapter-react-16";
import { configure, shallow } from "enzyme";
import examples from "./examples.data";
import { generateSnapshots } from "@microsoft/fast-jest-snapshots-react";
import { ICheckboxClassNameContract, ICheckboxState } from "./checkbox";
import {
    CheckboxProps,
    ICheckboxHandledProps,
    ICheckboxManagedClasses,
    ICheckboxUnhandledProps
} from "./checkbox.props";

/*
 * Configure Enzyme
 */
configure({adapter: new Adapter()});

describe("checkbox", (): void => {
    generateSnapshots(examples);
});

describe("checkbox unit-tests", (): void => {
    let Component: React.ComponentClass<ICheckboxHandledProps & ICheckboxManagedClasses>;
    let managedClasses: ICheckboxClassNameContract;

    beforeEach(() => {
        Component = examples.component;
        managedClasses = {
            checkbox: "checkbox-class",
            checkbox_disabled: "disabled-class",
            checkbox_input: "input-class",
            checkbox_label: "label-class",
            checkbox_span: "span-class",
        };
    });

    test("should correctly manage unhandledProps", () => {
        const handledProps: ICheckboxHandledProps & ICheckboxManagedClasses = {
            managedClasses
        };

        const unhandledProps: ICheckboxUnhandledProps = {
            "aria-hidden": true
        };

        const props: CheckboxProps = {...handledProps, ...unhandledProps};

        const rendered: any = shallow(
            <Component {...props} />
        );

        expect(rendered.props()["aria-hidden"]).not.toBe(undefined);
        expect(rendered.props()["aria-hidden"]).toEqual(true);
    });

    test("should correctly manage the disabled prop by adding a class and passing `disabled` to the input element", () => {
        const rendered: any = shallow(
            <Component managedClasses={managedClasses} disabled={true} />
        );

        expect(rendered.hasClass("disabled-class")).toBe(true);
        expect(rendered.find(".input-class[disabled]")).not.toBe(undefined);
        expect(rendered.find(".input-class[disabled]").prop("disabled")).toBe(true);
    });

    test("should correcly operate as an uncontrolled component and set initial state if `checked` prop is not passed", () => {
        const rendered: any = shallow(
            <Component managedClasses={managedClasses} />
        );

        const state: any = rendered.state("checked");

        expect(state).toBe(false);
    });

    test("should correcly operate as an uncontrolled component and handle `onChange` events when `onChange` prop is not passed", () => {
        const rendered: any = shallow(
            <Component managedClasses={managedClasses} />
        );

        expect(rendered.state("checked")).toBe(false);

        const input: any = rendered.find(".input-class");
        input.prop("onChange")(); // we have to fire the internal onChange from the input as we aren't passing one

        expect(rendered.state("checked")).toBe(true);
    });

    test("should correcly operate as a controlled component when `onChange` and `checked` prop is passed", () => {
        const onChange: any = jest.fn();
        const rendered: any = shallow(
            <Component managedClasses={managedClasses} checked={true} onChange={onChange} />
        );
        const input: any = rendered.find("input");

        input.simulate("change", { checked: false });

        expect(onChange).toHaveBeenCalled();
    });

    test("should correcly call getDerivedStateFromProps if an updated `checked` prop is passed which differs from the state", () => {
        const rendered: any = shallow(
            <Component managedClasses={managedClasses} />
        );

        const newProps: Partial<ICheckboxHandledProps> = {
            checked: true
        };

        expect(rendered.state("checked")).toEqual(false);

        rendered.setProps(newProps);

        expect(rendered.state("checked")).toEqual(true);
    });
});
