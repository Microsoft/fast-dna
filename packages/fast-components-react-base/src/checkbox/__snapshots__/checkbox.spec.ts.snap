// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox checkbox: 0 1`] = `
<label
  className="checkbox"
>
  <input
    checked={false}
    className="checkbox-input"
    disabled={null}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</label>
`;

exports[`checkbox checkbox: 1 1`] = `
<label
  className="checkbox"
>
  <input
    checked={false}
    className="checkbox-input"
    disabled={null}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</label>
`;

exports[`checkbox checkbox: 2 1`] = `
<div
  className="checkbox"
>
  <input
    checked={false}
    className="checkbox-input"
    disabled={null}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</div>
`;

exports[`checkbox checkbox: 3 1`] = `
<label
  className="checkbox"
>
  <input
    checked={false}
    className="checkbox-input"
    disabled={null}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</label>
`;

exports[`checkbox checkbox: 4 1`] = `
<label
  className="checkbox"
>
  <input
    checked={true}
    className="checkbox-input"
    disabled={null}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</label>
`;

exports[`checkbox checkbox: 5 1`] = `
<label
  className="checkbox"
>
  <input
    checked={true}
    className="checkbox-input"
    disabled={null}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</label>
`;

exports[`checkbox checkbox: 6 1`] = `
<label
  className="checkbox checkbox-disabled"
>
  <input
    checked={false}
    className="checkbox-input"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="checkbox-label"
  />
</label>
`;
